components/msg/background_hang_monitor.rs

```diff
@@ -0,0 +1,183 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use backtrace::Backtrace;
+use constellation_msg::{HangAlert, HangAnnotation};
+use constellation_msg::{MonitoredComponentId, MonitoredComponentMsg};
+use ipc_channel::ipc::IpcSender;
+use libc;
+use servo_channel::{Receiver, base_channel};
+use std::collections::HashMap;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::thread;
+use std::time::{Duration, Instant};
+
+
+/// The means of communication between monitored and monitor.
+pub static mut BACKTRACE: Option<(libc::pthread_t, Backtrace)> = None;
+
+lazy_static! {
+    /// A flag used to coordinate access to BACKTRACE.
+    pub static ref BACKTRACE_READY: AtomicBool = AtomicBool::new(false);
+}
+
+#[allow(unsafe_code)]
+#[cfg(not(windows))]
+unsafe fn get_backtrace_from_monitored_component(monitored: &MonitoredComponent) -> Option<Backtrace> {
+    libc::pthread_kill(monitored.thread_id, libc::SIGPROF);
+    loop {
+        if BACKTRACE_READY.load(Ordering::SeqCst) {
+            break;
+        }
+        thread::yield_now();
+    }
+    let (thread_id, trace) = BACKTRACE.take().unwrap();
+    assert_eq!(thread_id, monitored.thread_id);
+    BACKTRACE_READY.store(false, Ordering::SeqCst);
+    Some(trace)
+}
+
+#[cfg(target_os = "windows")]
+unsafe fn get_backtrace_from_monitored_component(monitored: &MonitoredComponent) -> Option<Backtrace> {
+    None
+}
+
+struct MonitoredComponent {
+    thread_id: libc::pthread_t,
+    last_activity: Instant,
+    last_annotation: Option<HangAnnotation>,
+    transient_hang_timeout: Duration,
+    permanent_hang_timeout: Duration,
+    sent_transient_alert: bool,
+    sent_permanent_alert: bool,
+    is_waiting: bool,
+}
+
+pub struct BackgroundHangMonitor {
+    monitored_components: HashMap<MonitoredComponentId, MonitoredComponent>,
+    constellation_chan: IpcSender<HangAlert>,
+    port: Receiver<(MonitoredComponentId, MonitoredComponentMsg)>,
+}
+
+impl BackgroundHangMonitor {
+    pub fn new(
+        constellation_chan: IpcSender<HangAlert>,
+        port: Receiver<(MonitoredComponentId, MonitoredComponentMsg)>,
+    ) -> Self {
+        BackgroundHangMonitor {
+            monitored_components: Default::default(),
+            constellation_chan,
+            port,
+        }
+    }
+
+    pub fn run(&mut self) -> bool {
+        let received = select! {
+            recv(self.port.select(), event) => {
+                match event {
+                    Some(msg) => Some(msg),
+                    // Our sender has been dropped, quit.
```

Comment:
I'm wondering if the assumption that the sender will be dropped is still valid, although I think it is. In single process mode even the constellation will own a sender, yet that will be dropped when the constellation exit, right?